HLA ={}
if 'DNASeq' in config:
	for subject in config['DNASeq']:
		for library in config['DNASeq'][subject]:
			if config['sample_captures'][library] not in config['Panel_List']:
				TARGET    += [subject+"/"+TIME+"/"+library+"/HLA/seq2HLA/"+library+"-ClassI.HLAgenotype4digits"]
				TARGET    += [subject+"/"+TIME+"/"+library+"/HLA/HLAminer/HLAminer_HPTASR.csv"]
				TARGET    += [subject+"/"+TIME+"/"+library+"/HLA/"+library+".Calls.txt"]
if 'matched_normal' in config:
	if len(config['matched_normal']) > 0:
		for Tumor in config['matched_normal']:
			for Normal in config['matched_normal'][Tumor]:
				seq2HLA    = "{subject}/{TIME}/{sample}/HLA/seq2HLA/{sample}-ClassI.HLAgenotype4digits".format(TIME=TIME, subject=SAMPLE_TO_SUBJECT[Normal], sample=Normal)
				HLAminer   = "{subject}/{TIME}/{sample}/HLA/HLAminer/HLAminer_HPTASR.csv".format(TIME=TIME, subject=SAMPLE_TO_SUBJECT[Normal], sample=Normal)
				if config['sample_captures'][Tumor] not in config['Panel_List']:
					# any output which is desired on all somatic libraries but Panel goes here, the list of panel captures should be maintained in the Panel_List in config file
					HLA[Tumor] = [seq2HLA, HLAminer]
	for sample in config['matched_normal'].keys():
		subject=SAMPLE_TO_SUBJECT[sample]
		if config['sample_captures'][sample] not in config['Panel_List']:
			TARGET    +=[subject+"/"+TIME+"/"+sample+"/NeoAntigen/MHC_Class_I/"+sample+".final.tsv"]



#if 'RNASeq' in config:
#	for subject  in config['RNASeq'].keys():
#		for sample in config['RNASeq'][subject]:
#			TARGET    +=  [subject+"/"+TIME+"/"+sample+"/HLA/seq2HLA/"+sample+"-ClassI.HLAgenotype4digits"]
#			TARGET    +=  [subject+"/"+TIME+"/"+sample+"/HLA/HLAminer/HLAminer_HPTASR.csv"]
#			TARGET    +=  [subject+"/"+TIME+"/"+sample+"/HLA/"+sample+".Calls.txt"]
#
############
#	seq2HLA
############
rule seq2hla:
	input:
		"{base}/{TIME}/FQ/{sample}_stat.txt",
		script=NGS_PIPELINE + "/seq2HLA/seq2HLA.py"
	output:
		"{base}/{TIME}/{sample}/HLA/seq2HLA/{sample}-ClassI.HLAgenotype4digits"
	params:
		rulename= "seq2hla",
		python  = config["version_python"],
		R	= config['version_R'],
		bowtie	= config['bowtie'],
		batch	= config[config['host']]["job_seq2hla"],
		HLA	= config['HLA']
	shell: """
	#######################
	module load bowtie/{params.bowtie} python/{params.python} R/{params.R}
	obj_get -v MoCha {wildcards.sample}_hg19_R1.fastq.gz --directory ${{LOCAL}}/
	obj_get -v MoCha {wildcards.sample}_hg19_R2.fastq.gz --directory ${{LOCAL}}/

	python {input.script} {params.HLA}/seq2HLA/ -1 ${{LOCAL}}/{wildcards.sample}_hg19_R1.fastq.gz -2 ${{LOCAL}}/{wildcards.sample}_hg19_R2.fastq.gz  -p ${{THREADS}} -r {wildcards.base}/{wildcards.TIME}/{wildcards.sample}/HLA/seq2HLA/{wildcards.sample}
	#######################
	"""
############
#	HLAminer
############
rule hla_miner:
	input:
		"{base}/{TIME}/FQ/{sample}_stat.txt",
		script=NGS_PIPELINE+"/HLAminer_v1.3.1/bin/HPTASRwgs_classI.sh"
	output: "{base}/{TIME}/{sample}/HLA/HLAminer/HLAminer_HPTASR.csv"
	params:
		rulename="hla_miner",
		batch	=config[config['host']]["job_annovar"],
		HLA	=config['HLA'],
        workdir=WORK_DIR,
		location=NGS_PIPELINE
	shell: """
	#######################
	obj_get -v MoCha {wildcards.sample}_hg19_R1.fastq.gz --directory ${{LOCAL}}/
	obj_get -v MoCha {wildcards.sample}_hg19_R2.fastq.gz --directory ${{LOCAL}}/
	echo ${{LOCAL}}/{wildcards.sample}_hg19_R1.fastq.gz >{wildcards.base}/{wildcards.TIME}/{wildcards.sample}/HLA/HLAminer/patient.fof
	echo ${{LOCAL}}/{wildcards.sample}_hg19_R2.fastq.gz >>{wildcards.base}/{wildcards.TIME}/{wildcards.sample}/HLA/HLAminer/patient.fof

	sh {input.script} {params.location}/HLAminer_v1.3.1/bin/ {wildcards.base}/{wildcards.TIME}/{wildcards.sample}/HLA/HLAminer/
	#######################
	"""
############
##       MergeHLA Calls
#############
rule combine_hla:
	input:
		A="{base}/{TIME}/{sample}/HLA/HLAminer/HLAminer_HPTASR.csv",
		B="{base}/{TIME}/{sample}/HLA/seq2HLA/{sample}-ClassI.HLAgenotype4digits",
	output:
		"{base}/{TIME}/{sample}/HLA/{sample}.Calls.txt"
	params:
		rulename = "combine_hla",
		Tool=NGS_PIPELINE + "/scripts/consensusHLA.pl",
		batch    = config[config['host']]["job_default"]
	shell: """
	#######################
	export LC_ALL=C
	perl {params.Tool} {input.B} {input.A} | sort > {output}	
	#######################
	"""
############
#	VEP4pVACSeq
############
rule vep:
	input:
		"{base}/{TIME}/{sample}/calls/{sample}.strelka.indels.raw.vcf",
		"{base}/{TIME}/{sample}/calls/{sample}.strelka.snvs.raw.vcf",
		"{base}/{TIME}/{sample}/calls/{sample}.MuTect.raw.vcf",
		HLA  =lambda wildcards: HLA[wildcards.sample],
		tool =NGS_PIPELINE + "/scripts/consensusSomaticVCF.pl",
	output: 
		vcf	="{base}/{TIME}/{sample}/NeoAntigen/{sample}.final.vcf",
	params:
		rulename = "vep",
		VEP	 = config['VEP'],
		merge=NGS_PIPELINE + "/scripts/consensusHLA.pl",
		normal	 = lambda wildcards: config['matched_normal'][wildcards.sample][0],
		batch    = config[config['host']]["job_VEP"],
	shell: """
	#######################
	module load vcftools VEP/{params.VEP} perl
	perl {input.tool} -vcf {wildcards.base}/{wildcards.TIME}/{wildcards.sample}/calls/{wildcards.sample}.strelka.indels.raw.vcf,{wildcards.base}/{wildcards.TIME}/{wildcards.sample}/calls/{wildcards.sample}.strelka.snvs.raw.vcf,{wildcards.base}/{wildcards.TIME}/{wildcards.sample}/calls/{wildcards.sample}.MuTect.raw.vcf -order {params.normal},{wildcards.sample} -filter REJECT |vcf-subset -u -c {wildcards.sample} >{output.vcf}.tmp
	variant_effect_predictor.pl -i {output.vcf}.tmp --plugin Downstream --plugin Wildtype --terms SO --offline --cache --dir_cache $VEPCACHEDIR --assembly GRCh37 --output_file {output.vcf} --vcf --force_overwrite
	rm -rf {output.vcf}.tmp
	export LC_ALL=C
	perl {params.merge} {input.HLA[0]} {input.HLA[1]} | sort > {wildcards.base}/{wildcards.TIME}/{params.normal}/HLA/{params.normal}.Calls.txt
	#######################
	"""
############
##       pVACSeq
#############
rule pvacseq:
	input:
		"{base}/{TIME}/{sample}/NeoAntigen/{sample}.final.vcf"
	output:
		"{base}/{TIME}/{sample}/NeoAntigen/MHC_Class_I/{sample}.final.tsv"
	version: config['pvacseq']
	params:
		rulename = "pvacseq",
		python   = config["version_python"],
		normal   = lambda wildcards: config['matched_normal'][wildcards.sample][0],
		tool 	 = NGS_PIPELINE + "/scripts/process_pVACSeq.pl",
		IEDB	 = config['IEDB'],
		batch    = config[config['host']]["job_VEP"],
		host	 = config['host']
	shell: """
	#######################
	allele=`cut -f1 {wildcards.base}/{wildcards.TIME}/{params.normal}/HLA/{params.normal}.Calls.txt |grep -v Allele|tr '\\n' ',' |sed -e 's/,$//g'`
	
	module load pvacseq/{version} python/{params.python}
	pvacseq run --iedb-install-directory {params.IEDB} -e 8  --fasta-size=200 {input} {wildcards.sample} ${{allele}} {{NNalign,NetMHC,NetMHCIIpan,NetMHCcons,NetMHCpan,PickPocket,SMM,SMMPMBEC,SMMalign}} {wildcards.base}/{wildcards.TIME}/{wildcards.sample}/NeoAntigen/8 &
	pvacseq run --iedb-install-directory {params.IEDB} -e 9  --fasta-size=200 {input} {wildcards.sample} ${{allele}} {{NNalign,NetMHC,NetMHCIIpan,NetMHCcons,NetMHCpan,PickPocket,SMM,SMMPMBEC,SMMalign}} {wildcards.base}/{wildcards.TIME}/{wildcards.sample}/NeoAntigen/9 &
	pvacseq run --iedb-install-directory {params.IEDB} -e 10 --fasta-size=200 {input} {wildcards.sample} ${{allele}} {{NNalign,NetMHC,NetMHCIIpan,NetMHCcons,NetMHCpan,PickPocket,SMM,SMMPMBEC,SMMalign}} {wildcards.base}/{wildcards.TIME}/{wildcards.sample}/NeoAntigen/10 &
	pvacseq run --iedb-install-directory {params.IEDB} -e 11 --fasta-size=200 {input} {wildcards.sample} ${{allele}} {{NNalign,NetMHC,NetMHCIIpan,NetMHCcons,NetMHCpan,PickPocket,SMM,SMMPMBEC,SMMalign}} {wildcards.base}/{wildcards.TIME}/{wildcards.sample}/NeoAntigen/11 &
	pvacseq run --iedb-install-directory {params.IEDB} -e 12 --fasta-size=200 {input} {wildcards.sample} ${{allele}} {{NNalign,NetMHC,NetMHCIIpan,NetMHCcons,NetMHCpan,PickPocket,SMM,SMMPMBEC,SMMalign}} {wildcards.base}/{wildcards.TIME}/{wildcards.sample}/NeoAntigen/12 &
	pvacseq run --iedb-install-directory {params.IEDB} -e 13 --fasta-size=200 {input} {wildcards.sample} ${{allele}} {{NNalign,NetMHC,NetMHCIIpan,NetMHCcons,NetMHCpan,PickPocket,SMM,SMMPMBEC,SMMalign}} {wildcards.base}/{wildcards.TIME}/{wildcards.sample}/NeoAntigen/13 &
	pvacseq run --iedb-install-directory {params.IEDB} -e 14 --fasta-size=200 {input} {wildcards.sample} ${{allele}} {{NNalign,NetMHC,NetMHCIIpan,NetMHCcons,NetMHCpan,PickPocket,SMM,SMMPMBEC,SMMalign}} {wildcards.base}/{wildcards.TIME}/{wildcards.sample}/NeoAntigen/14 &
	wait
	export LC_ALL=C
	cat {wildcards.base}/{wildcards.TIME}/{wildcards.sample}/NeoAntigen/*/*.final.tsv |sort |uniq >{output}
	perl {params.tool} {output} >{wildcards.base}/{wildcards.TIME}/{wildcards.sample}/NeoAntigen/{wildcards.sample}.final.txt
	#######################
	"""
