rule RNASeqQC:
	input:
		bam="{subject}/{TIME}/{sample}/{sample}.tophat.final.bam",
		bai="{subject}/{TIME}/{sample}/{sample}.tophat.final.bam.bai",
		rna_interval=config['rRNA_interval'],
		ref_flat=config['ref_flat'],
	output:
		table="{subject}/{TIME}/{sample}/qc/{sample}.RnaSeqMetrics.txt",
		pdf="{subject}/{TIME}/{sample}/qc/{sample}.RnaSeqMetrics.pdf",
	version: config["picard"]
	params:
		rulename  = "RnaSeqMetrics",
		R	  = config['version_R'],
		batch     = config[config['host']]["job_markdup"],
	shell: """
	#######################
	module load picard/{version}
	module load R/{params.R}
	java -Xmx${{MEM}}g -Djava.io.tmpdir=${{LOCAL}} -jar $PICARD_JAR CollectRnaSeqMetrics STRAND_SPECIFICITY=NONE VALIDATION_STRINGENCY=SILENT REF_FLAT={input.ref_flat} RIBOSOMAL_INTERVALS={input.rna_interval} INPUT={input.bam} OUTPUT={output.table} CHART_OUTPUT={output.pdf}
	#######################
	"""
rule RNASeqQC1:
	input:
		bam="{subject}/{TIME}/{sample}/{sample}.tophat.final.bam",
		bai="{subject}/{TIME}/{sample}/{sample}.tophat.final.bam.bai",
	output:
		table="{subject}/{TIME}/{sample}/qc/{sample}.AlignmentSummaryMetrics.txt",
	version: config["picard"]
	params:
		rulename  ="RnaSeqMetrics",
		batch     =config[config['host']]["job_markdup"],
		ref       =config['Bowtie2Index'].replace('genome', 'genome.fa')
	shell: """
	#######################
	module load picard/{version}
	java -Xmx${{MEM}}g -Djava.io.tmpdir=${{LOCAL}} -jar $PICARD_JAR CollectAlignmentSummaryMetrics VALIDATION_STRINGENCY=SILENT REFERENCE_SEQUENCE={params.ref} INPUT={input.bam} OUTPUT={output.table} ADAPTER_SEQUENCE=null
	#######################
	"""
rule RNASeqQC_1:
	input:
		"{subject}/{TIME}/{sample}/qc/fastqc/{sample}_R1_fastqc.html",
		file1="{subject}/{TIME}/{sample}/qc/{sample}.RnaSeqMetrics.txt",
		file2="{subject}/{TIME}/{sample}/qc/{sample}.AlignmentSummaryMetrics.txt",
		convertor=NGS_PIPELINE+ "/scripts/rnaseqQC.pl"	
	output:
		"{subject}/{TIME}/{sample}/qc/{sample}.RnaSeqQC.txt"
	params:
		rulename  = "RnaSeqMetrics",
		batch     = config[config['host']]["job_default"],
		source    = lambda wildcards: config['source'][wildcards.sample],
		diagnosis = lambda wildcards: config['diagnosis'][subject]
	shell: """
	#######################
	human="NA"
	if [[ {params.source} = 'PDX' ]]; then
		human=`grep hg19 {wildcards.subject}/{TIME}/FQ/{wildcards.sample}_stat.txt|cut -f2`
	fi	
	perl {input.convertor} {wildcards.subject}/{TIME}/{wildcards.sample}/qc/fastqc/{wildcards.sample}_R1_fastqc/fastqc_data.txt {input.file2} {input.file1} {wildcards.subject} {wildcards.sample} "{params.diagnosis}" ${{human}} >{output}
	#######################
	"""
rule RNASeqQC_2:
	input : files=lambda wildcards: SUB_QC[wildcards.subject],
		convertor=NGS_PIPELINE+"/scripts/transcript_coverage.R"
	output: 
		txt="{subject}/{TIME}/qc/{subject}.RnaSeqQC.txt",
		#plot="{subject}/{TIME}/qc/{subject}.transcriptCoverage.png"
	params:
		rulename  = "QC_Sum",
		batch     = config[config['host']]["job_covplot"]
	shell: """
	#######################
	export LC_ALL=C
	cat {input.files} |sort |uniq |sed -e '/^$/d'>{output.txt}
	#######################
	"""
rule RNASeqQC_3:
	input : RNA_QC_ALL
	output: "RnaSeqQC.txt"
	params:
		rulename  = "QC_Sum",
		batch     = config[config['host']]["job_default"]
	shell: """
	#######################
	touch {output}
	export LC_ALL=C
	cat {input} {output} |sort|uniq |sed -e '/^$/d'>{output}
	#######################
	"""
